GPU/High/HighGpu.cpp

```diff
@@ -0,0 +1,1048 @@
+}
+
+void HighGpuFrontend::ProcessEvent(GPUEvent ev) {
+	if (!backend_->ProcessEvent(ev)) {
+		GPUCommon::ProcessEvent(ev);
+	}
+}
+
+void HighGpuFrontend::PreExecuteOp(u32 op, u32 diff) { }
+
+void HighGpuFrontend::ExecuteOp(u32 op, u32 diff) {
+	const u8 cmd = op >> 24;
+	const CommandInfo info = cmdInfo_[cmd];
+	if (info.func) {
+		(this->*info.func)(op, diff);
+	}
+}
+
+void HighGpuFrontend::Execute_Vaddr(u32 op, u32 diff) {
+	gstate_c.vertexAddr = gstate_c.getRelativeAddress(op & 0x00FFFFFF);
+}
+
+void HighGpuFrontend::Execute_Iaddr(u32 op, u32 diff) {
+	gstate_c.indexAddr = gstate_c.getRelativeAddress(op & 0x00FFFFFF);
+}
+
+void HighGpuFrontend::Execute_Prim(u32 op, u32 diff) {
+	u32 data = op & 0xFFFFFF;
+	u32 count = data & 0xFFFF;
+	GEPrimitiveType prim = static_cast<GEPrimitiveType>(data >> 16);
+	if (count == 0)
+		return;
+
+	// Discard AA lines as we can't do anything that makes sense with these anyway. The SW plugin might, though.
+	if (gstate.isAntiAliasEnabled()) {
+		// Discard AA lines in DOA Paradise
+		if (prim == GE_PRIM_LINE_STRIP)
+			return;
+		// Discard AA lines in Summon Night 5
+		if ((prim == GE_PRIM_LINES) && gstate.isSkinningEnabled())
+			return;
+	}
+
+	if (gstate_c.skipDrawReason & (SKIPDRAW_SKIPFRAME | SKIPDRAW_NON_DISPLAYED_FB))	{
+		// TODO: Figure out a way to do cycle estimates here, without disturbing the draw engine
+		// which might be running on another thread.
+		return;
+	}
+
+	if (!Memory::IsValidAddress(gstate_c.vertexAddr)) {
+		ERROR_LOG_REPORT(G3D, "Bad vertex address %08x!", gstate_c.vertexAddr);
+		return;
+	}
+
+	bool indexed = false;
+	if ((gstate.vertType & GE_VTYPE_IDX_MASK) != GE_VTYPE_IDX_NONE) {
+		if (!Memory::IsValidAddress(gstate_c.indexAddr)) {
+			ERROR_LOG_REPORT(G3D, "Bad index address %08x!", gstate_c.indexAddr);
+			return;
+		}
+		indexed = true;
+	}
+
+	dirty_ = CommandSubmitDraw(cmdPacket_, &arena_, &gstate, dirty_, data, gstate_c.vertexAddr, indexed ? gstate_c.indexAddr : 0);
+
+	// After submitting the drawcall, we advance the vertexAddr (when non indexed) or indexAddr (when indexed).
+	// Some games rely on this, they don't bother reloading VADDR and IADDR.
+	// The VADDR/IADDR registers are NOT updated.
+	if (indexed) {
+		if ((gstate.vertType & GE_VTYPE_IDX_MASK) == GE_VTYPE_IDX_16BIT) {
+			count *= sizeof(u16);
+		}
+		gstate_c.indexAddr += count;
+	} else {
+		// TODO: Add a very small (4-entry?) LRU cache for these. Cannot use the vertexdecodercache here
+		// as it belongs to the drawing thread, and it's not worth adding locking.
+		int vertexSize = ComputePSPVertexSize(gstate.vertType);
+		gstate_c.vertexAddr += count * vertexSize;
+	}
+
+	if (cmdPacket_->full)
+		FlushCommandPacket();
+}
+
+void HighGpuFrontend::Execute_Bezier(u32 op, u32 diff) {
+	// TODO
+}
+
+void HighGpuFrontend::Execute_Spline(u32 op, u32 diff) {
+	// TODO
+}
+
+void HighGpuFrontend::Execute_BlockTransferStart(u32 op, u32 diff) {
+	CommandSubmitTransfer(cmdPacket_, &gstate);
```

Comment:
May need to flush if full afterward, right?  Could definitely happen, I think I saw a sequence of block transfers in some games.

-[Unknown]